lib/Dialect/Arc/Transforms/LowerStateRewrite.cpp

```diff
@@ -0,0 +1,1206 @@
+    auto ifResetOp = createOrReuseIf(module.builder, loweredReset, true);
+    module.builder.setInsertionPoint(ifResetOp.thenYield());
+
+    // Generate the zero value writes.
+    for (auto state : states) {
+      auto type = cast<StateType>(state.getType()).getType();
+      Value value = module.builder.create<ConstantOp>(
+          loweredReset.getLoc(),
+          module.builder.getIntegerType(hw::getBitWidth(type)), 0);
+      if (value.getType() != type)
+        value = module.builder.create<BitcastOp>(loweredReset.getLoc(), type,
+                                                 value);
+      module.builder.create<StateWriteOp>(loweredReset.getLoc(), state, value,
+                                          Value{});
+    }
+    module.builder.setInsertionPoint(ifResetOp.elseYield());
+  }
+
+  // Handle the enable.
+  if (enable) {
+    // Check if we can reuse a previous enable value.
+    auto &[unloweredEnable, loweredEnable] = module.prevEnable;
+    if (unloweredEnable != enable ||
+        loweredEnable.getParentBlock() != module.builder.getBlock()) {
+      unloweredEnable = enable;
+      loweredEnable = lowerValue(enable, Phase::Old);
+      if (!loweredEnable)
+        return failure();
+    }
+
+    // Check if we're inserting right after an if op for the same enable, in
+    // which case we can reuse that op. Otherwise create the new if op.
+    auto ifEnableOp = createOrReuseIf(module.builder, loweredEnable, false);
+    module.builder.setInsertionPoint(ifEnableOp.thenYield());
+  }
+
+  // Get the transfer function inputs. This potentially inserts read ops.
+  SmallVector<Value> loweredInputs;
+  for (auto input : inputs) {
+    auto lowered = lowerValue(input, Phase::Old);
+    if (!lowered)
+      return failure();
+    loweredInputs.push_back(lowered);
+  }
+
+  // Compute the transfer function and write its results to the state's storage.
+  auto loweredResults = createMapping(loweredInputs);
+  for (auto [state, value] : llvm::zip(states, loweredResults))
+    module.builder.create<StateWriteOp>(value.getLoc(), state, value, Value{});
+
+  // Since we just wrote the new state value to storage, insert read ops just
+  // before the if op that keep the old value around for any later ops that
+  // still need it.
+  module.builder.setInsertionPoint(ifClockOp);
+  for (auto [state, result] : llvm::zip(states, results)) {
+    auto oldValue = module.builder.create<StateReadOp>(result.getLoc(), state);
+    module.loweredValues[{result, Phase::Old}] = oldValue;
+  }
+
+  return success();
+}
+
+/// Lower a memory and its read and write ports to corresponding
+/// `arc.memory_write` operations. Reads are also executed at this point and
+/// stored in `loweredValues` for later operations to pick up.
+LogicalResult OpLowering::lower(MemoryOp op) {
+  if (phase != Phase::New)
+    return op.emitOpError() << "cannot be lowered in the " << phase << " phase";
```

Comment:
Can these errors actually be triggered from some input IR or only if this function is used incorrectly? If the latter, we could make it an assert. If the earlier, maybe some error test?

Reply:
These should never be triggered by user input. This was just a sanity check while developing the pass, where it was useful to get a meaningful error message. I can definitely turn these into asserts.